R/data.table.R

```diff
@@ -1942,9 +1942,32 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       X[[j]] = xj
     }
   }
-  X = unlist(X, recursive = FALSE, use.names = FALSE)
-  dim(X) <- c(n, length(X)/n)
-  dimnames(X) <- list(rownames.value, unlist(collabs, use.names = FALSE))
+  
+  # convert columns to common class before handing over to C
+  col.classes = unlist(lapply(X[-rownames], class))
+  class.order = c("logical"=1L, "integer"=2L, "numeric"=3L, "complex"=4L,
+                  "character"=5L, "raw"=6L, "list"=7L)
+  if (length(unique(col.classes)) > 1L) { 
+    target.class = names(which.max(class.order[unique(col.classes)]))
+    if (target.class == "raw") target.class = "character" # to match behaviour of as.matrix.data.frame
+    for (col in which(col.classes != target.class)) {
+      ifelse(col >= rownames, col + 1L, col)
+      switch(target.class,
+             # 'logical' can never be reached, it will always be converted to something more complex 
+             "integer" = { X[[col]] = as.integer(X[[col]]) }, 
+             "numeric" = { X[[col]] = as.numeric(X[[col]]) }, 
+             "complex" = { X[[col]] = as.complex(X[[col]]) }, 
+             "character" = { X[[col]] = as.character(X[[col]]) }
+             # 'raw' can never be reached - columns should all be converted to 'character' if any were 'raw'
+             # 'list' can never be reached - if there is one list column, all will converted to list columns already
+      )
+    }
+  }
+  
+  p = length(X[-rownames])
+  X = .Call(Casmatrix, X, n, p, rownames)
```

Comment:
I wonder if the signature is too complicated, why can't we just pass `X` and `rownames`?

Reply:
The signature has been shortened as suggested; I needed to determine n and p in the C code to make sure they don't exceed 2^31 - 1 elements (i.e. by bitshifting in C rather than computing the 2^31 - 1 in R and comparing)